src/components/Tool/FixMailingAddresses/FixMailingAddresses.tsx

```diff
@@ -150,17 +150,53 @@ const FixMailingAddresses: React.FC<Props> = ({ accountListId }: Props) => {
       });
     });
     setSourceOptions([...existingSources]);
+
+    const newDataState = data
+      ? data.contacts.nodes?.reduce(
+          (map, contact) => ({
+            ...map,
+            [contact.id]: {
+              addresses: contact.addresses.nodes.map((address) => {
+                existingSources.add(address.source);
+                return { ...address };
+              }),
+            },
+          }),
+          {},
+        )
+      : {};
+
+    setDataState(newDataState);
   }, [loading, data]);
 
+  const handleChangePrimary = (
+    contactId: string,
+    numberIndex: number,
+  ): void => {
+    if (!dataState[contactId]) {
+      return;
+    }
+
+    const temp = { ...dataState };
+
+    temp[contactId].addresses = temp[contactId].addresses.map(
+      (address, index) => ({
+        ...address,
+        primaryMailingAddress: index === numberIndex,
+      }),
+    );
+    setDataState(temp);
+  };
+
   const handleSingleConfirm = async ({
```

Comment:
Wrap this in `useCallback` as it is passed down. This also goes for any function that is passed down to another component.